pkg/torcx/propagate.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 CoreOS Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package torcx
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+)
+
+const ()
+
+// propagateUnits installs all systemd unit and unit-like files as transient
+// units in /run/systemd.
+// Units are taken from /usr/lib/systemd
+func propagateUnits(applyCfg *ApplyConfig, imageRoot string) error {
+
+	srcDir := filepath.Join(imageRoot, "usr", "lib", "systemd")
+	st, err := os.Stat(srcDir)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	if !st.IsDir() {
+		return nil
+	}
+
+	_, err = os.Stat(SYSTEMD_DIR)
+	if err != nil {
+		return errors.Wrapf(err, "error checking for systemd volatile directory %s", SYSTEMD_DIR)
+	}
+
+	// Walk the source directory
+	err = filepath.Walk(srcDir, func(srcPath string, srcStat os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relPath, err := filepath.Rel(srcDir, srcPath)
+		if err != nil {
+			return err
+		}
+		if relPath == "." {
+			return nil
+		}
+
+		dstPath := filepath.Join(SYSTEMD_DIR, relPath)
+		dstStat, dstStatErr := os.Stat(dstPath)
+		if dstStatErr != nil && !os.IsNotExist(dstStatErr) {
+			return err
+		}
+
+		// Copy src to dest
+		// * If src is dir, mkdir if not exists
+		// * If src is symlink, duplicate link contents exactly
+		// * If src is file, symlink
+		if srcStat.IsDir() {
+			if os.IsNotExist(err) {
+				return os.Mkdir(dstPath, 0755)
```

Comment:
Is there any reason to create empty directories? Why not use MkdirAll when placing files instead?